### DISCUSSION TITLE
Loading initrds takes a long time

### DISCUSSION THREAD

**Discussion by freym** [ipxe/ipxe#264] [Q&A] [answered]
Hello

I am trying out IPXE to replace syslinux.
I have noticed that the boot process is much slower. When I boot with syslinux, it downloads the files and starts right away with booting.
IPXE seems to download the initrd and then do something with it. And this process takes up to 30 seconds or more, even with smaller initrds. And with big ones it can take up to 5 minutes or more.
Is there any way to speed up this process or at least see what it does  ?

IPXE Version: 1.21.1+ (g19d0)
Kernel: http://ftp.debian.org/debian/dists/buster/main/installer-amd64/current/images/netboot/debian-installer/amd64/linux
Initrd: http://ftp.debian.org/debian/dists/buster/main/installer-amd64/current/images/netboot/debian-installer/amd64/initrd.gz

The download is done with HTTP.
Sample script:
```
#!ipxe

imgfetch /initrd.gz
kernel /linux initrd=initrd.gz
boot linux
```

**Comment by pbartfai** [comment]
I have the same experience under UEFI, but it works better with BIOS boot.

**Comment by freym** [comment]
I have tested it on multiple Machines/VMs with UEFI or Bios. I do not see any differences.

**Comment by mcb30** [comment]
Just tested using a KVM virtual machine with those exact files.  It takes about five seconds from VM power-on to reaching the Debian language selection screen.  Same for both BIOS and UEFI.

Could you share more details about the environment in which you are able to reproduce the delay?

**Reply by mcb30** [reply]
@freym Here's a sample five-second boot:

https://user-images.githubusercontent.com/692245/108208639-30b59c80-7121-11eb-9b1e-9785c61f26e8.mp4

**Comment by freym** [comment]
Ok I found the problem:
We are using a Python application to generate the IPXE scripts and serve the static files. If we use a real webserver like Nginx to serve the static files, everything works with no delay. But this only happens when IPXE downloads files, it does not happen when I download the files with syslinux or grub2.

**Reply by mcb30** [reply]
Good find.  It sounds as though there may be something erroneous in the HTTP protocol implementation.  Could you capture a packet trace (https://ipxe.org/howto/pcap) from the server side?

**Reply by mcb30** [reply]
Alternatively, could you provide a trivial Python script that will reproduce the problem?  (Assuming that you're using one of the standard web-serving frameworks.)

**Comment by freym** [comment]
The application is Vinegar a open source boot server:
https://github.com/KIT-IBPT/vinegar

They are using the http.server library from Python3.
https://docs.python.org/3/library/http.server.html

I will try to capture a packet trace,

**Reply by mcb30** [reply]
Thanks.  I tested using `http.server` to serve the files via the usual static one-liner:
```
python -m http.server 8000
````
Booted fine in just a few seconds, so this does look to be a bug in https://github.com/KIT-IBPT/vinegar

**Comment by mcb30** [comment, marked as answer]
The problem turns out to be that the server software (https://github.com/KIT-IBPT/vinegar) does not provide the HTTP Content-Length header.  This causes iPXE to continually reallocate memory for every incoming packet, since it has no idea how much memory needs to be allocated to hold the file.

The Content-Length header (or equivalent information, such as the use of chunked transfer encoding) is mandatory as of HTTP/1.1.  The server software in question supports only HTTP/1.0, where the header is not mandatory but is strongly advised.

I have created https://github.com/KIT-IBPT/vinegar/pull/3 to fix the issue in the server software.

**Reply by mcb30** [reply]
@freym My PR https://github.com/KIT-IBPT/vinegar/pull/3 has been merged: if you use the latest upstream then you should find the problem has now gone away.
